pkg/libovsdb/schema.go

```diff
@@ -71,3 +79,344 @@ func (schema DatabaseSchema) validateOperations(operations ...Operation) bool {
+		if err := json.Unmarshal(*colTypeJSON.ValueRawMsg, &column.TypeObj.Value.Type); err != nil {
+			if err := json.Unmarshal(*colTypeJSON.ValueRawMsg, &column.TypeObj.Value); err != nil {
+				return fmt.Errorf("cannot parse value object: %s", err)
+			}
+			if err := column.TypeObj.Value.parseEnum(*colTypeJSON.ValueRawMsg); err != nil {
+				return err
+			}
+		}
+		if !isAtomicType(column.TypeObj.Value.Type) {
+			return fmt.Errorf("schema contains unknown atomic type %s", column.TypeObj.Key.Type)
+		}
+	}
+
+	// Technially, we have finished unmarshalling. But let's finish infering the native
+	if column.TypeObj.Value != nil {
+		column.Type = TypeMap
+	} else if column.TypeObj.Min != 1 || column.TypeObj.Max != 1 {
+		column.Type = TypeSet
+	} else if len(column.TypeObj.Key.Enum) > 0 {
+		column.Type = TypeEnum
+	} else {
+		column.Type = column.TypeObj.Key.Type
+	}
+	return nil
+}
+
+// parseEnum decodes the enum field and populates the BaseType.Enum field
+func (bt *BaseType) parseEnum(rawData json.RawMessage) error {
+	// EnumJSON is used to dynamically decode the Enum values
+	type EnumJSON struct {
+		Enum interface{} `json:"enum,omitempty"`
+	}
+	var enumJSON EnumJSON
+
+	if err := json.Unmarshal(rawData, &enumJSON); err != nil {
+		return fmt.Errorf("cannot parse enum object: %s (%s)", string(rawData), err)
+	}
+	// enum is optional
+	if enumJSON.Enum == nil {
+		return nil
+	}
+
+	// 'enum' is a list or a single element representing a list of exactly one element
+	switch enumJSON.Enum.(type) {
+	case []interface{}:
+		// it's an OvsSet
+		oSet := enumJSON.Enum.([]interface{})
+		innerSet := oSet[1].([]interface{})
+		bt.Enum = make([]interface{}, len(innerSet))
+		for k, val := range innerSet {
+			bt.Enum[k] = val
+		}
+	default:
+		bt.Enum = []interface{}{enumJSON.Enum}
+	}
+	return nil
+}
+
+func isAtomicType(atype string) bool {
+	switch atype {
+	case TypeInteger, TypeReal, TypeBoolean, TypeString, TypeUUID:
+		return true
+	default:
+		return false
+	}
+}
+
+func (columnSchema ColumnSchema) AtomicTypeDefault() interface{} {
+	switch columnSchema.Type {
+	case TypeInteger:
+		return int(0)
+	case TypeReal:
+		return float64(0)
+	case TypeBoolean:
+		return false
+	case TypeString:
+		return ""
+	case TypeUUID:
+		return UUID{"00000000-0000-0000-0000-000000000000"}
+	default:
+		panic(fmt.Sprintf("Unsupported type %s", columnSchema.Type))
+	}
+}
+
+func (tableSchema TableSchema) Default(row *map[string]interface{}) {
+	for column, columnSchema := range tableSchema.Columns {
+		if _, ok := (*row)[column]; !ok {
+			(*row)[column] = columnSchema.AtomicTypeDefault()
+		}
+	}
+}
+
+func (databaseSchema DatabaseSchema) Default(table string, row *map[string]interface{}) {
+	tableSchema, ok := databaseSchema.Tables[table]
+	if !ok {
+		panic(fmt.Sprintf("Missing schema for table %s", table))
+	}
+	tableSchema.Default(row)
+}
+
+func (schema Schema) Default(dbname, table string, row *map[string]interface{}) {
```

Comment:
it is not part of the schema processing

Reply:
as explained above, its an integral part of managing schemas for multiple databases. Renamed to `Schemas` as a greed